compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp

```diff
@@ -691,12 +692,157 @@ FailureOr<Flow::WorkloadBuilder> getWorkloadBuilder(OpBuilder &builder,
+  // d4, d6, d7).
+  ReassociationIndices range;
+  for (auto [idx, dim] : llvm::enumerate(loopDims)) {
+    if (!range.empty()) {
+      if (!areSameDimExprForAllAffineMaps(dim) || (range.back() + 1 != dim)) {
+        if (range.size() > 1)
+          contiguousLoops.push_back({range.begin(), range.end()});
+        range.clear();
+      }
+    }
+    if (areSameDimExprForAllAffineMaps(dim)) range.push_back(dim);
+  }
+  if (range.size() > 1) contiguousLoops.push_back(range);
+
+  // Convert dimensions to positions
+  for (auto &loop : contiguousLoops)
+    for (auto &idx : loop) idx = firstMap.getDimPosition(idx);
+  return contiguousLoops;
+}
+
+/// Finds all collapsible loop ranges of type "parallel" without
+/// mixing them with "reduction" types.
+static SmallVector<ReassociationIndices> getCollapsibleLoops(
+    linalg::GenericOp genericOp) {
+  SmallVector<unsigned int> pDims;
+  genericOp.getParallelDims(pDims);
+
+  SmallVector<ReassociationIndices> pLoops =
+      getContiguousLoops(genericOp, pDims);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Collapsing parallel loops: ";
+    for (auto indices : pLoops) {
+      llvm::dbgs() << "[";
+      for (auto idx : indices) llvm::dbgs() << idx << ",";
+      llvm::dbgs() << "]\t";
+    }
+    llvm::dbgs() << "\n";
+  });
+
+  return pLoops;
+}
+
+/// Collapse possible dimension of the given linalg.generic and return the new
+/// one
+static FailureOr<linalg::GenericOp> collapseLinalgGeneric(
+    TensorDimTrackingRewriter &rewriter, linalg::GenericOp genericOp) {
+  SmallVector<ReassociationIndices> collapseIndices =
+      getCollapsibleLoops(genericOp);
+
+  rewriter.setInsertionPoint(genericOp);
+  FailureOr<SmallVector<Value>> replacements =
+      mlir::linalg::collapseGenericOpIterationDims(genericOp, collapseIndices,
+                                                   rewriter);
+  if (failed(replacements) || replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
+
+  // Find and return collapsed linalg.generic
+  if (auto expandshapeOp =
+          replacements->front().getDefiningOp<tensor::ExpandShapeOp>()) {
+    if (auto newGenericOp =
+            expandshapeOp.getOperand().getDefiningOp<linalg::GenericOp>()) {
+      rewriter.replaceOp(genericOp, *replacements);
+      return newGenericOp;
+    }
+  }
+
+  return failure();
+}
+
+static constexpr unsigned kNumMaxParallelDims = 3;
+
+/// Returns true if the given op is collapsable.
+static bool isEligibleForCollapse(Operation *op,
+                                  ArrayRef<Operation *> producers) {
+  if (!producers.empty()) return false;
+
+  auto genericOp = dyn_cast<linalg::GenericOp>(op);
+  if (!genericOp) return false;
+
+  // TODO There is no mechanism to convey collapsed indexes from
+  // `tensor.collapse_shape` to `tensor.expand_shape`. Once we have this support
+  // in MLIR, we can enable dynamic tensor shapes.
```

Comment:
nit we need a `:` right after `TODO`.

```suggestion
  // TODO: There is no mechanism to convey collapsed indexes from
  // `tensor.collapse_shape` to `tensor.expand_shape`. Once we have this support
  // in MLIR, we can enable dynamic tensor shapes.
```

Maybe we don't have a convention in IREE. IMO, you can add your account name to TODO, which lets others know you're the best person who understands the logic and reason.